session.go

```diff
@@ -333,12 +334,12 @@ runLoop:
 		s.handshakeChan <- handshakeEvent{err: closeErr.err}
 	}
 	s.handleCloseError(closeErr)
-	close(s.runClosed)
+	s.ctxCancel()
```

Comment:
I know we shouldn't return earlier, but I'd feel safer if this was deferred at the beginning of the func.